### PR TITLE
HHH-16298 Failure merging a referenced entity 

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16298

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 1]
```diff
@@ -460,14 +465,33 @@ else if ( value == null ) {
...
-			EntityPersister entityPersister = getAssociatedEntityPersister( session.getFactory() );
-			Object propertyValue = entityPersister.getPropertyValue( value, uniqueKeyPropertyName );
+			if ( value instanceof HibernateProxy ) {
+			/*
+				If the value is a Proxy and the property access is field, the value returned by
+			 	`attributeMapping.getAttributeMetadata().getPropertyAccess().getGetter().get( object )`
+			 	is always null except for the id, we need the to use the proxy implementation to
+			 	extract the property value.
+			 */
+				value = ( (HibernateProxy) value ).getHibernateLazyInitializer().getImplementation();
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:475]:
```suggestion
			final LazyInitializer lazyInitializer = extractLazyInitializer( value );
			if ( lazyInitializer != null ) {
				/*
				If the value is a Proxy and the property access is field, the value returned by
			 	`attributeMapping.getAttributeMetadata().getPropertyAccess().getGetter().get( object )`
			 	is always null except for the id, we need the to use the proxy implementation to
			 	extract the property value.
				*/
				value = lazyInitializer.getImplementation();
```

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java [hunk 2]
```diff
@@ -460,14 +465,33 @@ else if ( value == null ) {
...
+			}
+			else if ( ManagedTypeHelper.isPersistentAttributeInterceptable( value ) ) {
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/type/EntityType.java:477]:
```suggestion
			else if ( isPersistentAttributeInterceptable( value ) ) {
```

**Review by mbladel — COMMENTED** [review]
Other than some slight refactoring LGTM.

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/type/EntityType.java:475]: We usually use `extractLazyInitializer` to handled cases like this.

File: hibernate-core/src/main/java/org/hibernate/type/EntityType.java (hunk 2 above)
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/type/EntityType.java:477]: Import static `isPersistentAttributeInterceptable` as well.

**Review by beikov — APPROVED** [review]

**Pull request merged**